### ISSUE TITLE
Snowman never chooses Microsoft64 calling convention

### ISSUE THREAD

**Issue by yegord** [yegord/snowman#27]
One should either add some algorithm choosing between AMD64 and Microsoft64 calling convention, or give the user the ability to choose a convention manually and redecompile everything.

**Comment by yegord** [comment]
https://github.com/yegord/snowman/blob/master/src/nc/arch/x86/X86MasterAnalyzer.cpp#L175

**Comment by hlide** [comment]
ok, I'm checking "microsoft64" right now

**Comment by hlide** [comment]
here a sequence in WIN64 ABI:
```
1400013c0:	push rdi
1400013c2:	sub rsp, 0x20
1400013c6:	mov rdi, rsp
1400013c9:	mov ecx, 0x8
1400013ce:	mov eax, 0xcccccccc
1400013d3:	rep stosd 
1400013d5:	lea rcx, [rip+0xdfac]
1400013dc:	call dword 0x40001186
1400013e1:	lea rcx, [rip+0x8888]
1400013e8:	call dword 0x40004b80
1400013ed:	add rsp, 0x20
1400013f1:	pop rdi
1400013f2:	ret 
```

the result with "microsoft64":

```
int64_t fun_1400013c0() {
    int32_t ecx1;
    int64_t* rax2;
    int64_t rax3;

    ecx1 = 8;
    while (ecx1) {
        --ecx1;
    }
    rax2 = fun_140001186(0x14000f388);
    rax3 = fun_140004b80(fun_140009c70);
    return rax3;
}
```

it sounds correct, except for two things:

1) ``` ecx1 = 8; while (ecx1) { --ecx1; } ``` either it should be complete or it should be discarded entirely.

2) *rax2* should have been discarded.

**Comment by hlide** [comment]
Assembly code:
```
1400024a0:	mov [rsp+0x10], rdx
1400024a5:	mov [rsp+0x8], ecx
1400024a9:	push rdi
1400024aa:	sub rsp, 0x40
1400024ae:	mov rdi, rsp
1400024b1:	mov ecx, 0x10
1400024b6:	mov eax, 0xcccccccc
1400024bb:	rep stosd 
1400024bd:	mov ecx, [rsp+0x50]
1400024c1:	mov dword [rsp+0x20], 0xffffffff
1400024c9:	lea rcx, [rip+0x9950]
1400024d0:	call qword [rip+0xeb32]
1400024d6:	mov [rsp+0x28], rax
1400024db:	cmp qword [rsp+0x28], 0x0
1400024e1:	jz 0x14000250d
1400024e3:	lea rdx, [rip+0x994e]
1400024ea:	mov rcx, [rsp+0x28]
1400024ef:	call qword [rip+0xeb0b]
1400024f5:	mov [rsp+0x30], rax
1400024fa:	mov rdx, [rsp+0x58]
1400024ff:	mov ecx, [rsp+0x50]
140002503:	call qword [rsp+0x30]
140002507:	mov [rsp+0x20], eax
14000250b:	jmp 0x14000252f
14000250d:	call qword [rip+0xed9d]
140002513:	mov ecx, 0x30
140002518:	imul rcx, rcx, 0x2
14000251c:	add rax, rcx
14000251f:	lea rdx, [rip+0x991a]
140002526:	mov rcx, rax
140002529:	call qword [rip+0xece1]
14000252f:	mov eax, [rsp+0x20]
140002533:	add rsp, 0x40
140002537:	pop rdi
140002538:	ret 
```

snowman c++:
```
int64_t LoadLibraryA = 0x11832;

int64_t __iob_func = 0x11a00;

int64_t fprintf = 0x11a18;

int32_t fun_14000115e(int32_t ecx, int64_t rdx, int64_t r8) {
    int64_t v4;
    int32_t v5;
    int32_t ecx6;
    int32_t v7;
    int64_t rax8;
    int64_t v9;
    int64_t rax10;
    int64_t rax11;
    int64_t rcx12;
    int32_t eax13;

    v4 = rdx;
    v5 = ecx;
    ecx6 = 16;
    while (ecx6) {
        --ecx6;
    }
    v7 = -1;
    rax8 = (int64_t)LoadLibraryA("pspe4all-emud.dll");
    v9 = rax8;
    if (v9 == 0) {
        rax10 = (int64_t)__iob_func("pspe4all-emud.dll");
        fprintf(rax10 + 96, "Cannot load pspe4all-emu.dll.");
    } else {
        rax11 = (int64_t)GetProcAddress(v9, "Run");
        *(int32_t*)&rcx12 = v5;
        *((int32_t*)&rcx12 + 1) = 0;
        eax13 = (int32_t)v9(rcx12, v4);
        v7 = eax13;
    }
    return v7;
}
```

real source:
```
int _tmain(int argc, _TCHAR* argv[])
{
    int result = -1;

    HMODULE hEmuDll = LoadLibraryA(__ConfigurationDllStringA("pspe4all-emu"));

    if (hEmuDll)
    {
        auto Run = (int(*)(int, wchar_t *[]))GetProcAddress(hEmuDll, "Run");

        result = Run(argc, argv);
    }
    else
    {
        fprintf(stderr, "Cannot load pspe4all-emu.dll.");
    }

    return result;
}
```

**Comment by hlide** [comment]
that part:
```
        rax11 = (int64_t)GetProcAddress(v9, "Run");
        *(int32_t*)&rcx12 = v5;
        *((int32_t*)&rcx12 + 1) = 0;
        eax13 = (int32_t)v9(rcx12, v4);
        v7 = eax13;
```
is wrong and makes no sense

**Comment by yegord** [comment]
> 2) rax2 should have been discarded.

Could you check if this patch helps: d56d5d54d5f389a92d39153572806b2c2c52d7d8?

**Comment by hlide** [comment]
```
int64_t fun_1400013c0() {
    int32_t ecx1;
    int64_t rax2;

    ecx1 = 8;
    while (ecx1) {
        --ecx1;
    }
    fun_140001186(0x14000f388);
    rax2 = fun_140004b80(fun_140009c70);
    return rax2;
}
```

**Comment by hlide** [comment]
something is really wrong:

``` 
140005bd0:	cmp ecx, 0x4
140005bd3:	ja 0x140005bea
140005bd5:	movsxd rax, ecx
140005bd8:	lea r8, [rip+0x9429]
140005bdf:	mov ecx, [r8+rax*4] <--- it swaps between old value
140005be3:	mov [r8+rax*4], edx <--- and new value so it is important to keep this order
140005be7:	mov eax, ecx <-- return old value
140005be9:	ret 
140005bea:	or eax, 0xffffffff
140005bed:	ret 
```

the decompiling result makes no sense:
```
int64_t fun_140005bd0(uint32_t ecx, int32_t edx) {
    int64_t rax3;
    int64_t rax4;

    if (ecx > 4) {
        return 0xffffffff;
    } else {
        rax3 = (int64_t)(int32_t)ecx;
        *(int32_t*)((int64_t)"" + rax3 * 4) = edx; <------ what !? old value is discarded !?
        *(int32_t*)&rax4 = *(int32_t*)((int64_t)"" + rax3 * 4); <------ what's that !?
        *((int32_t*)&rax4 + 1) = 0; <------ what's that !?
        return rax4;
    }
}
```

**Comment by yegord** [comment]
> *((int32_t*)&rax4 + 1) = 0; <------ what's that !?

This is easy: implicit zero extend. :-)

> *(int32_t*)&rax4 = *(int32_t*)((int64_t)"�" + rax3 * 4); <------ what's that !?

r8 is a constant, the decompiler wrongly guesses that r8 is a pointer to a C string, and inserts that string there.

> old value is discarded !?

Is likely a bug related to https://github.com/yegord/snowman/blob/master/src/nc/core/ir/cgen/DefinitionGenerator.cpp#L1171

Yeah, I have plenty of work to do...

**Comment by hlide** [comment]
well I did understand it was the zero extending but why not:
```
int32_t ecx5;

ecx5 = *(int32_t*)((int64_t)"�" + rax3 * 4);
*(int32_t*)((int64_t)"�" + rax3 * 4) = edx;
rax4 = (int64_t)ecx5;
return rax4;
```
doing `*((int32_t*)&reg) = value;` instead of `reg = value;` sounds clumsy.

**Comment by yegord** [comment]
> why not:

Snowman is stupid: if in IR there is an assignment to lower half of rax, and an assignment to the higher half of rax, it generates two assignments.

**Comment by yegord** [comment]
Now microsoft64 convention is chosen automatically for whatever looks like a Window binary: 6b22d5604855f1c607372a9245c6be079a7ace9b.

**Comment by nihilus** [comment]
Wouldnt the PE-parser tell us that?

**Comment by nihilus** [comment]
Actually based on the PE-parser it is easy to write a Unix COFF-parser

**Comment by nihilus** [comment]
Ive got fsn.COFF, famous file navigator, from first Jurassic Park. Yes it is MIPS :-)

**Comment by nihilus** [comment]
I will add some test-cases to that Repo during weekend

**Comment by yegord** [comment]
Yes, PE parser sets the operating system to Windows, which is then taken into account by detectCallingConvention().

**Comment by nihilus** [comment]
It could determine if it is x86 or not as well. I added MIPS-support to it as well.

**Comment by nihilus** [comment]
You know NT 4.0 was released for MIPS and Alpha machines... And on WinCE MIPS survived another 10-15 years.

**Comment by nihilus** [comment]
Actually matching against COFF in the IDA plugin is a huge potential mistake for x86 UNIX COFF binaries from SCO, Linux and Xenix etc.

**Comment by nihilus** [comment]
Actually the assumption based on that COFF is Windows is not true at all.

**Comment by yegord** [comment]
nihilius@ If you want to be useful, submit a patch with a better OS detection algorithm.

**Comment by nihilus** [comment]
Well let me first implement a parser for fsn.COFF :-)
I aint complaining just giving you my thoughts of the observation. I will ask ppl I know about the PE-format, might even go ask my old friend bda@microsoft.com  (Bruce Dang) for some help.

**Comment by hlide** [comment]
@nihilus just avoid to add those email address in public, bots can scan them and spam your friend.

**Comment by nihilus** [comment]
@hlide Good point, point taken. :)

**Comment by nihilus** [comment]
@yegord My patch would be     ```if (inf.filetype == f_WIN || inf.filetype == f_PE) {``` instead of ```    if (inf.filetype == f_WIN || inf.filetype == f_COFF || inf.filetype == f_PE) {``` really simple ;-)

**Comment by yegord** [comment]
Sorry, but I would like to see an evidence that majority of COFF files are not built for Windows. The question comes from the fact that I never saw a UNIX COFF file in wild life: I saw a.out (FreeBSD 2.2.8!) and ELF.

After all, PE does not mean 100% Windows either: it could be, say, an EFI module. But for some reason you do not complain about that.

**Comment by nihilus** [comment]
Yepp, could be EFI-based that is true... :-( I dont know if there is a way to get the PE or COFF header in IDA Pro SDK. From their one could see the magic. Look for linux/coff.h for the Linux 386 magic. here is another COFF: ftp://ftp.sgi.com/sgi/fsn/fsn.COFF.tar.Z

**Comment by nihilus** [comment]
Oh, fun thing is that IDA want to load that one as Intel 960 COFF.

**Comment by yegord** [comment]
Ok, we now choose Microsoft64 convention for Windows binaries (6b22d5604855f1c607372a9245c6be079a7ace9b).
The expression reconstruction for the swap example should be also fixed (0b4770cef2897e9ebbb2171b3e8e6d0647430ef1, 81011a8125ae15363f68d7fe54aa197e1a738b65).
Absolute garbage should not be recognized as ASCII strings either (55e9e52422c1a80794a03a99312682b32c1dbf29).
